File: 6주차/LeetCode_128_Longest_Consecutive_Sequence/dongjun.java
/**
 *  @문제접근
 *  HashSet을 이용해 풀었습니다.
 *  nums 배열의 모든 값을 set에 저장한 후,
 *  각 숫자에서 연속적으로 몇개의 숫자가 set에 저장되어 있는지
 *  카운팅하여, 가장 높은 카운트를 반환하게 구현했습니다.
 *
 *  @성능
 *  Runtime: 3 ms
 *  Memory Usage: 39.9 MB
 *  시간복잡도: O(N)
 */
public class LongestConsecutiveSequence {
    public int longestConsecutive(int[] nums) {
        if(nums.length == 0)
            return 0;
        int cnt = 1;
        int answer = 1;
        HashSet<Integer> set = new HashSet<>();

        for(int n: nums)
            set.add(n);

        for(int n: nums){
            if(!set.contains(n-1)){
                int temp = n;
                while(set.contains(++temp))
                    cnt++;

                answer = Math.max(answer, cnt);
                cnt = 1;
            }
        }
        return answer;
    }
}

File: 8주차/LeetCode_152_Maximum_Product_Subarray/dongjun.py
"""
1. 문제접근
	동적계획법을 사용해 풀었습니다.
	각각의 최솟값과 최댓값을 모두 저장해
	음수 * 음수의 경우를 처리했습니다.

2. 성능
	Runtime: 60ms
	Memory Usage: 16.2MB
	시간복잡도: O(N)
	공간복잡도: O(N)
"""

class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        MIN, MAX = 0, 1
        dp = [[1]*2 for _ in range(len(nums))]
        
        dp[0][MIN] = nums[0]
        dp[0][MAX] = nums[0]
        answer = nums[0]
        
        for i in range(1, len(nums)):
            dp[i][MIN] = min(dp[i-1][MAX]*nums[i], dp[i-1][MIN]*nums[i], nums[i])
            dp[i][MAX] = max(dp[i-1][MAX]*nums[i], dp[i-1][MIN]*nums[i], nums[i])
            
            answer = max(dp[i][MAX], answer)
        
        return answer
